### ISSUE TITLE
Possible bug with Safari on OSX (tested with Safari 8.0.5 on OSX 10.10.3)

### ISSUE THREAD

**Issue by mariusk** [rackt/react-router#907]
I'm using a pattern in my app (which starts from "/"):

```javascript
    if (!router.goBack()) {
      router.transitionTo('/');
    }
```

My app starts in "/", I press a link which then goes to "/add", where I press a cancel button that invokes the code above (call `goBack` first, and to root if it fails). On Chrome this works fine. On the platforms specified in the subject it does not. I am taken to the "/add" route, but when I press cancel the screen does not move.

I've narrowed it down and found a difference in the window.history state in `History.js` inside the `back` function. If I output `window.history` to the console before calling `window.history.back()`, on Chrome (where it is working), `window.history` is the expected "/".  On Safari when it does not work it contains "/add" (which points to the current page so nothing happens).

Unfortunately I do not have a minimial test case for this (yet).

**Comment by mariusk** [comment]
Further digging, it seems when my discardButton handler which invokes the code pasted above gets clicked, the following happens on Safari:

```
AddMatch::_onClickDiscard
bundle.js:454WINDOW HISTORY {"length":46,"state":{"path":"/"}}
bundle.js:454WINDOW HISTORY {"length":46,"state":{"path":"/add"}}
```

On Chrome, the last line which goes **back** to the /add route never happens. Not sure why this is happening on Safari  yet.

**Comment by mariusk** [comment]
So here's a weird thing; In this function from History.js:

```javascript
  /**
   * Sends the browser back one entry in the history.
   */
  back: function () {
    invariant(
      canUseDOM,
      'Cannot use History.back without a DOM'
    );

    // Do this first so that History.length will
    // be accurate in location change listeners.
    History.length -= 1;

    window.history.back();
  }
```
If I comment out the call to `window.history.back()`, everything works fine, both in Chrome and in Safari. That means I'm not sure what this code is effectively meant to do. It does not seem to do any harm on Chrome, but it *does* seem to effectively make the back call get called twice on Safari.

**Comment by gaearon** [comment]
Let me be silly here.. How does it go back (and pop state) if you comment out `history.back()` call?

**Comment by mariusk** [comment]
Yeah, that was bugging me as well. I figured there must be more code paths than I had identified. As is often the case, this bug has nothing to do with react-router or Safari, and everything to do with my POS code.

**Comment by mariusk** [comment]
It turns out I'm trying to do fancy stuff with tracking routes as part of app state (in an "AppStateStore"). Here's a link that has some magic code in it showing one way to do this, using some kind of proxy to avoid circular dependencies: https://github.com/rackt/react-router/blob/master/docs/guides/flux.md#accessing-route-and-params-from-action-creators

So I've got a couple of Actions (flux style) in my app which are named "routeTo" (which eventually calls router transitionTo) and "goBack" (which calls router goBack). Related to me getting my foot in the mouth is the fact that I have recently started transitioning most of my coding from Coffeescript to ES6. Coffeescript has "implicit" return from functions, javascript does not. Here's what my "router proxy" code (see link above) had before I figured it out:

```javascript
  goBack() {
    router.goBack();
  },
```

Does that return a value? On Chrome, it looks like it. On Safari not. This is after running everything through babel of course. Adding and *explicit* returns fixes everything. Since I had a direct call to transitionTo if goBack fails, I guess that is what tricked me to thinking it seems to work.

Oh well, live and learn!
